src/library_syscall.js

```diff
@@ -209,21 +209,55 @@ var SyscallsLibrary = {
 #else
     var stream = SYSCALLS.getStreamFromFD(fd);
     switch (op) {
-      case {{{ cDefs.TCGETA }}}:
+      case {{{ cDefs.TCGETA }}}: {
+        if (!stream.tty) return -{{{ cDefs.ENOTTY }}};
+#if SYSCALL_DEBUG
+        dbg('warning: not filling tio struct');
+#endif
+        return 0;
+      }
       case {{{ cDefs.TCGETS }}}: {
         if (!stream.tty) return -{{{ cDefs.ENOTTY }}};
+        if (stream.tty.ops.ioctl_tcgets) {
+          var termios = stream.tty.ops.ioctl_tcgets(stream);
+          var argp = SYSCALLS.get();
+          {{{ makeSetValue('argp', C_STRUCTS.termios.c_iflag, 'termios.c_iflag || 0', 'i32') }}};
+          {{{ makeSetValue('argp', C_STRUCTS.termios.c_oflag, 'termios.c_oflag || 0', 'i32') }}};
+          {{{ makeSetValue('argp', C_STRUCTS.termios.c_cflag, 'termios.c_cflag || 0', 'i32') }}};
+          {{{ makeSetValue('argp', C_STRUCTS.termios.c_lflag, 'termios.c_lflag || 0', 'i32') }}};
+          for (var i = 0; i < {{{ cDefs.NCCS }}}; i++) {
+            {{{ makeSetValue('argp + i', C_STRUCTS.termios.c_cc, 'termios.c_cc[i] || 0', 'i8') }}};
+          }
+          return 0;
+        }
 #if SYSCALL_DEBUG
         dbg('warning: not filling tio struct');
 #endif
         return 0;
       }
       case {{{ cDefs.TCSETA }}}:
       case {{{ cDefs.TCSETAW }}}:
-      case {{{ cDefs.TCSETAF }}}:
+      case {{{ cDefs.TCSETAF }}}: {
+        if (!stream.tty) return -{{{ cDefs.ENOTTY }}};
+        return 0; // no-op, not actually adjusting terminal settings
+      }
       case {{{ cDefs.TCSETS }}}:
       case {{{ cDefs.TCSETSW }}}:
       case {{{ cDefs.TCSETSF }}}: {
         if (!stream.tty) return -{{{ cDefs.ENOTTY }}};
+        if (stream.tty.ops.ioctl_tcsets) {
+          var optional_actions = op == {{{ cDefs.TCSETS }}} ? "" : op == {{{ cDefs.TCSETSW }}} ? "W" : "F";
```

Comment:
It is an argument to distinguish the event types: `""` stands for `TCSETS`, `"W"` does for `TCSETSW`, and `"F"` does for `TCSETSF`.

I am open to change the values. Do you like to pass the numbers like `cDefs.TCSETS` or `cDefs.TCSETSW`? Or pass a string like `"TCSETS"` or `"TCSETSW"`?

Reply:
Maybe just stick with `cDefs.TCSETS` since they already exist?

Reply:
Maybe just stick with `cDefs.TCSETS` since they already exist?

Reply:
I did so, thanks! 16fff9d5175bca1741aa5fca35e06518fb761e71

Reply:
Just FYI: The reason why I did something a little different is that the constants such as `cDefs.TCSETS` cannot be referred from xterm-pty which is outside of Emscripten. But, the values of the constants probably won't change easily, so it's not a big deal. I think we should aim to incorporate xterm-pty into Emscripten in the end.

Reply:
Right that makes sense.

Reply:
> Just FYI: The reason why I did something a little different is that the constants such as `cDefs.TCSETS` cannot be referred from xterm-pty which is outside of Emscripten. But, the values of the constants probably won't change easily, so it's not a big deal. I think we should aim to incorporate xterm-pty into Emscripten in the end.

BYW if you write the glue code between xterm.js and emscripten as a JS library file you can access all the macros.  e.g. `--js-library=library_xterm.js` .. can code than interpret the emscripten requests and make calls into the external xterm.js library.

Reply:
Yes, I would like to challenge that next.

However, I think it may be difficult to get everything working nice with just `--js-library=library_xterm.js`.

This is because `getchar` function needs to synchronously wait for user input. The default implementation of `getchar` waits synchornously by using `prompt` function, but we can't do the same for xterm.js. So, xterm-pty requires to run the Emscripten runtime in a Web Worker, and waits for user input via SharedArrayBuffer.

Therefore, `emcc` needs to generate two JS files, one for the main thread and one for the Web Worker. I wonder if this is beyond the scope of the `--js-library=library_xterm.js` option.

Reply:
We have an option called PROXY_TO_THREAD that puts the main application code in worker.   When `getchar` is called from a worker it should be able to block..   We also have `-sASYNCIFY` and `-sASYNCIFY=2` (A.K.A JSPI) which allow functions to suspend/block, even when running on the main thread.

Reply:
Great to know PROXY_TO_THREAD, thanks! I will read the code.

> We also have `-sASYNCIFY` and `-sASYNCIFY=2` (A.K.A JSPI) which allow functions to suspend/block, even when running on the main thread.

I'm not sure about this. As long as `getchar` is not an async function, I thought it could not wait synchronously in the main thread. Maybe I need to learn JSPI.

Reply:
`-sASYNCIFY` and `-sASYNCIFY=2` allow the wasm module to suspend its computation and resume later.  For example, after some async work is done.  From the POV of the code being suspended it appears like a blocking call.